### DISCUSSION TITLE
64-bit support für yf-akcarea-host

### DISCUSSION THREAD

**Discussion by fda77** [Freetz-NG/freetz-ng#852] [Allgemein]
Das Tool für die [avm-kernel-config area](https://github.com/Freetz-NG/freetz-ng/tree/master/make/host-tools/yf-akcarea-host/) von Peterpawn das er13 erweiterte ist nur als 32bit binary ausführbar, das hatte ich vor längerem irgendwo mit @PeterPawn herausgefunden. Daran Schuld dürften Dinge wie [swapEndianness](https://github.com/Freetz-NG/freetz-ng/blob/master/make/host-tools/yf-akcarea-host/src/lib_avm_kernel_config.c#L30) sein. Ich kann das selbst nicht beheben, vielleicht hast du irgend wann mal nichts besseres zu tun @er13, oder auch sonst wer der sich damit auskennt. Zum testen eignet sich zb eine 7490.

Dass das Tool nur als 32bit ausführbar ist ist bei x86_64 soweit noch egal, bei aarch64 (arm 64bit) aber nicht das diese arm nur 64bit code compilieren & ausführen können. Außer mit einem 32bit Betriebsystem, dann ist aber auch nur 32bit Code ausführen. Ein flotten aarch64 gibt es zb von Oracle als kostenlosen VPS mit 4 cores und 24gb ram und 45gb hdd... Mit solch einem konnte ich durch ein paar kleinere Änderungen ein (7590) Image compilieren: 4dacc35ee a414a024b e10bb9931 

Dadurch dass yf-akcarea-host nicht nutzbar ist können mit aarch64 aktuell nur v2 und v4 Kernel (Module) gebaut wurden. Von den v3 Kernel nur solche bei denen die [FREETZ_AVM_KERNEL_CONFIG_AREA_KNOWN](https://github.com/Freetz-NG/freetz-ng/blob/master/config/avm/kernel.in#L526) unbekannt ist. ("unbekannt" bedeutet hier dass die Automatik des Tools diese Grösse nicht ermitteln kann, das könnte auch noch angepasst werden). Komplett ausgenommen sind v4 Kernel, für diese würde jemand der replace-kernel benötigt wohl besser ein separates tool erstekkeb. Avm hat es mittlweile geschafft Codeschnipsel dazu zu veröffentlichen.
Falls yf-akcarea-host mal 64bit beherrschen sollte dürften für fakeroot&pseudo die biarch Abschnitte ganz entfernt werden können.

Zum Basteln:
 - Symbol `FREETZ_HOSTTOOLS_DOWNLOAD` ausschalten
 - Symbol `FREETZ_REPLACE_KERNEL` einer Kernel v3 Box (zb 7490) einschalten
 - Diese Zeilen entfernen:
   * fd2556424 yf-akcarea-host is 32bit only
   * c02bf0ecd enable aarch64 except when yf-akcarea-host is used by v3 kernel
 - `make libdtc-host-recompile`
 - Source in https://github.com/Freetz-NG/freetz-ng/tree/master/make/host-tools/yf-akcarea-host/src
 - `make yf-akcarea-host-recompile`
 - später zum testen `make kernel-recompile`

**Comment by er13** [comment]
@fda77 

Mit swapEndianness hat es nichts zu tun, sondern schlicht und ergreifend damit, dass alle Fritz!Boxen eine 32-Bit CPU haben bzw. diese im 32-Bit Modus betrieben wird (weiß nicht, ob bei irgendwelchen ARM-basierten Fritz!Boxen die CPU auch 64-Bit unterstützt). Kernel-Config-Area ist also eine Struktur/ein Array in der Bitness der Target-CPU => also 32-Bit. Die einfachste Art und Weise, diese Struktur zu parsen ist es auf dem Host in derselben Bitness zu arbeiten wie die der Target-CPU. Arbeitet man in derselben Bitness, muss man nur die Endianness swappen.

Möchte man nun auf dem Host in der von der Target-CPU abweichenden Bitness arbeiten, so muss man beim Parsen der Kernel-Config-Area neben swapEndianness noch die 32-Bit Adressen/Zahlen in 64-Bit Adressen/Zahlen umwandeln und beim Rausschreiben der .asm-Datei dann wieder, da wo notwendig, die 64-Bit Adressen/Zahlen in 32-Bit zurück umwandeln.

**Comment by er13** [comment]
Ein weiteres Problem ist die Tatsache, dass es versucht wird, entweder die originale oder eine auf dieser basierte Version von avm_kernel_config.h zu verwenden, s. https://github.com/Freetz-NG/freetz-ng/blob/master/make/host-tools/yf-akcarea-host/src/avm_kernel_config.h

Diese definiert Strukturen, die Pointer enthalten, und sizeof(pointer) auf 32-Bit ist nun mal nicht identisch mit sizeof(pointer) auf 64-Bit.

D.h. man muss avm_kernel_config.h anpassen (=gleichbedeutend die aus dem AVM-open-source-Paket kann definitiv nicht mehr verwendet werden) und alle Pointer in den Strukturen durch uint32 ersetzen (Größe der Adressen auf einer 32-Bit-CPU), und an den Stellen, wo auf diese Pointer bisher zugegriffen wurde, eine weitere Indirektionsschicht einführen - gelesen wird eine 32-Bit unsigned int, auf dem Host ist diese in eine 64-Bit Adresse umzuwandeln und dann zu dereferenzieren. Der Code wird dadurch hässlich und kaum lesbar.

**Comment by PeterPawn** [comment]
Das ganze Tool hat sich überlebt, seitdem seitens AVM im Open-Source-Paket auch die Dateien veröffentlicht werden, mit denen man den Quelltext für diese Struktur im Kernel gleich direkt erzeugen kann - zu finden unterhalb von `sources/kernel/kgw` in den neuen AVM-Paketen (zumindest in dem für die 7590).

Meines Erachtens reitet man also einen toten Gaul, wenn man weitere Arbeit in dieses Tool steckt - aber ich wiederhole mich gerne (jedoch auch zum letzten Mal in diesem Kontext).

**Reply by er13** [reply]
Danke für den Hinweis (bin seit mindestens 2 Jahren nicht mehr aktiv und habe gar nicht in die letzten Open-Source-Paketen geschaut).


>  mit denen man den Quelltext für diese Struktur im Kernel gleich direkt erzeugen kann - zu finden unterhalb von sources/kernel/kgw in den neuen AVM-Paketen

Wenn das wirklich so ist, dann muss man in der Tat kein Aufwand in `yf-akcarea` mehr stecken.

**Comment by fda77** [comment]
Prima, das macht die Sache wesentlich einfacher! Die aarch64 Probleme sind weg und die noch fehlenden Kernel sollten damit funktionieren. In die Sourcepakete von Avm schau ich so gut wie nie rein, das die 5mb Patches für Freetz von einem Script erstellt werden. Solange es durchläuft ists gut.

Ich hab mit die 7490 (kernel v3) + 7590(kernel v4) + 7520 (arm) 7.56 Sourcen herausgepickt, bei allen 3 sind die Dateien in `sources/kernel/kgw/scripts/` identisch. Der Einfachheit halber gepackt: [akc-7490+7590+7520_7.56.zip](https://github.com/Freetz-NG/freetz-ng/files/12704802/akc-7490%2B7590%2B7520_7.56.zip)
Avms `"Config & Symbol Maschine"` bastelt das ganze mit perl, python und bash zusammen. Scheinbar wird der Kernel normal gebaut und erst danach eine Binärdatei compiliert (kernel config area) die dann mittels `dd` chirurgisch in den Kernel eingesetzt wird.
Am witzigsten ist die Stelle mit der Ermittlung des Bereiches in [link_and_print_symbols.sh .txt](https://github.com/Freetz-NG/freetz-ng/files/12704843/link_and_print_symbols.sh.txt)  :)
```
	config_start=`grep __avm_kernel_config_start ${KERNEL_DIR}/boot/System.map`
	config_end=`grep __avm_kernel_config_end ${KERNEL_DIR}/boot/System.map`
```

Ich denke für Freetz sollte das ganz am Ende von fwmod ausgeführt werden, wenn auch die Kernelmodule kopiert werden. Denn theoretisch müssen in die Area alle Module die ins Image kommen.
Praktisch ist die die Area (glaub ich jedenfalls) aktuell in Freetz je SOURCE_ID immer gleich

**Reply by fda77** [reply]
Die Area mit den Werkzeugen von AVM zu bauen ist keine gute Idee solang man 1 Source für mehr als 1 Geräte verwendet.
Als Beispiel der dvb-c und 1750. Der dvb hat max FOS 7.0x, 1750 7.3x. Von der Hardware hat der dvb-c "nur 1 usb hub mit 2 Tunern" mehr. Mit 1750 Software als Alien auf der DVBC Hardware und es funktioniert kein Wlan.

Vergleich Fos 7.0 von dvbc und 1750
```
--- 1759-70.dts
+++ 1750-70.dts
@@ -4,13 +4,12 @@
        compatible = "qca,ath79";
        #address-cells = <0x01>;
        #size-cells = <0x01>;
-       model = "AVM FritzRepeater DVB-C";
+       model = "AVM FritzRepeater 1750E";

        aliases {
                spi0 = "/ahb/spi@1f000000";
                serial0 = "/ahb/apb/uart@18020000";
                pci0 = "/ahb/apb/pci@0x14000000";
-               ehci_0 = "/ahb/apb/usb@1b000000";
                eth0 = "/ahb/ethernet@19000000";
                eth1 = "/ahb/ethernet@1a000000";
        };
@@ -111,8 +110,8 @@
                                #interrupt-cells = <0x01>;
                                interrupt-parent = <0x01>;
                                interrupts = <0x02>;
-                               linux,phandle = <0x06>;
-                               phandle = <0x06>;
+                               linux,phandle = <0x05>;
+                               phandle = <0x05>;
                        };

                        interrupt-controller@ip3_int {
@@ -121,8 +120,6 @@
                                #interrupt-cells = <0x01>;
                                interrupt-parent = <0x01>;
                                interrupts = <0x03>;
-                               linux,phandle = <0x05>;
-                               phandle = <0x05>;
                        };

                        reset-controller@1806001c {
@@ -145,26 +142,14 @@
                                i2c-rx-fifo-depth = <0x40>;
                                clocks = <0x03 0x03>;
                                clock-names = "ref";
-                               status = "okay";
-                       };
-
-                       usb@1b000000 {
-                               compatible = "qca,ath79-ehci\0qca,ap135-ehci";
-                               reg = <0x1b000000 0x1000>;
-                               interrupt-parent = <0x05>;
-                               interrupts = <0x00>;
-                               has-transaction-translator;
-                               qca_force_host_mode;
-                               qca_force_16bit_ptw;
-                               caps-offset = <0x100>;
-                               status = "okay";
+                               status = "disabled";
                        };

                        pci@0x14000000 {
                                compatible = "qca,ar724x-pci";
                                reg = <0x14000000 0x1000 0x180f0000 0x100 0x180c0000 0x1000 0x10000000 0x2000000 0x00 0x01>;
                                reg-names = "cfg_base\0ctrl_base\0crp_base\0mem_base\0io_base";
-                               interrupt-parent = <0x06>;
+                               interrupt-parent = <0x05>;
                                interrupts = <0x00>;
                                status = "okay";
                        };
@@ -190,8 +175,8 @@
                        ar934x-support = <0x00>;
                        reset-bit = <0x400000>;
                        status = "disabled";
-                       linux,phandle = <0x07>;
-                       phandle = <0x07>;
+                       linux,phandle = <0x06>;
+                       phandle = <0x06>;

                        ethernet-phy@0 {
                                reg = <0x00>;
@@ -202,7 +187,7 @@
                        compatible = "qca,ag71xx-eth";
                        reg = <0x19000000 0x200 0x18070000 0x14>;
                        reg-names = "mac_base\0cfg_base";
-                       mdio-handle = <0x07>;
+                       mdio-handle = <0x06>;
                        eth-pll-data = <0x1313 0x00 0x86000000>;
                        phy-mode = <0x03>;
                        force-speed = <0x3e8>;
@@ -226,7 +211,7 @@
                        compatible = "qca,ag71xx-eth";
                        reg = <0x1a000000 0x200>;
                        reg-names = "mac_base";
-                       mdio-handle = <0x07>;
+                       mdio-handle = <0x06>;
                        eth-pll-data = <0x1616 0x101 0x3000101>;
                        phy-mode = <0x07>;
                        force-speed = <0x3e8>;
@@ -272,104 +257,85 @@
                };

                gpio_avm_led_fsi1 {
-                       value = <0x68>;
+                       value = <0x65>;
                };

                gpio_avm_led_fsi2 {
-                       value = <0x67>;
+                       value = <0x66>;
                };

                gpio_avm_led_fsi3 {
-                       value = <0x66>;
+                       value = <0x67>;
                };

                gpio_avm_led_fsi4 {
-                       value = <0x65>;
-               };
-
-               gpio_avm_led_tv {
-                       value = <0x69>;
+                       value = <0x68>;
                };

-               gpio_avm_led_power {
-                       value = <0x6a>;
+               gpio_avm_led_lan {
+                       value = <0x0d>;
                };

                gpio_avm_led_wlan {
-                       value = <0x6b>;
+                       value = <0x69>;
                };

-               gpio_avm_em28174_reset {
-                       value = <0x6c>;
+               shift_register_size {
+                       value = <0x08>;
+                       param = <0x0d>;
+                       function = <0x00>;
                };

-               shift_register_size {
-                       value = <0x10>;
+               gpio_avm_led_power {
+                       value = <0x6a>;
+                       default = <0x01>;
                };

                gpio_avm_button_wlan {
                        value = <0x04>;
-                       param = <0x0c>;
+                       param = <0x0d>;
                };

                gpio_avm_uart_rx {
                        value = <0x09>;
+                       param = <0x0d>;
+                       function = <0x08>;
                };

                gpio_avm_uart_tx {
                        value = <0x0a>;
+                       param = <0x0b>;
+                       function = <0x16>;
                };

                gpio_avm_extphy1_reset {
                        value = <0x0b>;
-                       param = <0x0a>;
+                       param = <0x0b>;
                        default = <0x01>;
                };

                gpio_avm_mdio_clk {
                        value = <0x0c>;
+                       param = <0x0b>;
+                       function = <0x21>;
                };

                gpio_avm_mdio_data {
                        value = <0x13>;
+                       param = <0x11>;
+                       function = <0x200310>;
                };

                gpio_avm_peregrine_reset {
-                       value = <0x6d>;
+                       value = <0x11>;
+                       param = <0x0b>;
+                       default = <0x01>;
                };

                gpio_avm_pcie_reset {
-                       value = <0x6e>;
-               };
-
-               gpio_avm_spi_clk {
-                       value = <0x0e>;
-               };
-
-               gpio_avm_spi_data {
-                       value = <0x0f>;
-               };
-
-               gpio_avm_mdio_irq {
-                       value = <0x10>;
-               };
-
-               gpio_avm_i2c_clk {
                        value = <0x12>;
-                       param = <0x10>;
-                       out-function = <0x07>;
-                       in-function = <0x12>;
-               };
-
-               gpio_avm_i2c_data {
-                       value = <0x11>;
-                       param = <0x10>;
-                       out-function = <0x06>;
-                       in-function = <0x13>;
-               };
-
-               usb_hub_available {
-                       value = <0x01>;
+                       param = <0x0b>;
+                       default = <0x01>;
                };
        };

@@ -385,7 +351,7 @@
                compatible = "qca,wifi-ar955x";
                reg = <0xb8100000 0x20000>;
                core-id = <0x00>;
-               interrupt-parent = <0x06>;
+               interrupt-parent = <0x05>;
                interrupts = <0x01>;
                interrupt-names = "legacy";
                status = "ok";
@@ -397,7 +363,7 @@

        avm-hw-revision {
                compatible = "avm,avm_hw_revision";
-               revision = "205";
+               revision = "206";
                subrevision = "0";
        };
 };
```

Dh man muss immer die Area (bzw den DeviceTree darin) passend zu genau der Hardware nehmen, auch je FOS sind die verschieden.
Man könnte nun aus build/original/ die Area des Images (nicht Sourcen) nehmen und dies in den replaec-kernel wie avm mit dd einbauen. Dies könnte man vorher auspacken und die Module um die von Freetz erweitern. 

Nur leider wäre für das Extrahieren weiterhin das akc-host tool notwendig das kein 64bit kann...
(Für Alien FWs müsste man sogar noch die Möglichkeit einbauen eine abweichende dts einzubauen)
Man könnte in Freetz auch eine dts-Datenbank aus dem AVM-Images generieren, Openwrt hat das auch so: https://github.com/openwrt/openwrt/blob/main/target/linux/ath79/dts/qca9556_avm_fritz1750e.dts

Ich weiss jetzt nicht so recht ob das alles so eine gute Idee wäre, vielleicht hat jemand bessere
